### DISCUSSION TITLE
error -1001 at ggml-opencl.cpp:965

### DISCUSSION THREAD

**Discussion by SpeedyCraftah** [ggml-org/llama.cpp#2316] [Q&A]
I've followed the build guide for CLBlast in the README - I've installed `opencl-headers` and compiled OpenCL from source as well as CLBlast and then built the whole thing with cmake.

When I tried to run `./main`, I get this weird assertion error:
```
llamacpp$ ./main
main: build = 874 (7d5f184)
main: seed  = 1689986093
ggml_opencl: clGetPlatformIDs(NPLAT, platform_ids, &n_platforms) error -1001 at ggml-opencl.cpp:965
```
Any ideas?
Thank you!

OS - WSL2 Ubuntu 22.04 with CUDA support (confirmed CUDA support with pytorch)
GPU - Nvidia GeForce RTX 3080

**Comment by ghost** [comment]
Hi, the error means llama.cpp tried to use OpenCL on your machine, but it failed.

`ocl-icd` will likely link the libaries for you.

**Reply by SpeedyCraftah** [reply]
Thanks!

`E: Unable to locate package ocl-icd`
I also tried to run it but to no avail.

I do have `ocl-icd-opencl-dev` installed though if that helps?

**Reply by ghost** [reply]
> I do have `ocl-icd-opencl-dev` installed though if that helps?
>
I'm not familiar with Ubuntu, so if you already tried that then OpenCL still isn't linked as expected.

I would try 1. `git clone https://github.com/OCL-dev/ocl-icd`

2.  Compile then Install `ocl-icd`: https://github.com/OCL-dev/ocl-icd/blob/master/INSTALL#L58

This should be enough to enable CLBlast through OpenCL.

One way to test OpenCL is correctly linked is with `clinfo`.

**Reply by SpeedyCraftah** [reply]
Thank you!
I don't actually think the issue is related to the linking of OpenCL - I ran the `clinfo` command as you mentioned and got this:
```
llama$ clinfo
Number of platforms                               0
```
I had a look at the vendors and Nvidia is in there:
```
llama$ cat /etc/OpenCL/vendors/*
libnvidia-opencl.so.1
```

I did see someone mention this may be due to a version mismatch of Cuda versions and my versions do not appear to match up?
```
llama$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Thu_Nov_18_09:45:30_PST_2021
Cuda compilation tools, release 11.5, V11.5.119
Build cuda_11.5.r11.5/compiler.30672275_0

llama$ nvidia-smi
Tue Jul 25 13:31:25 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.54.06              Driver Version: 536.40       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3080        On  | 00000000:2B:00.0  On |                  N/A |
|  0%   39C    P8              24W / 340W |   1446MiB / 10240MiB |      6%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A        23      G   /Xwayland                                 N/A      |
+---------------------------------------------------------------------------------------+
```
I'm not quite sure what the issue is - I even downloaded Cuda WSL from the official Nvidia website that says it is Cuda 2.0, but they appear mismatched, not sure if the versions shown here are supposed to match.
Anyhow I will try again this time using an Ubuntu VM with GPU passthrough and hopefully have some luck there, if not I'll also try putting Ubuntu on a flash drive and attempting to set things up there.

**Reply by SpeedyCraftah** [reply]
On a side note I did also try setting things up on Windows before realising NVCC isn't supported on Windows (what a shame).

**Reply by ghost** [reply]
The fact `clinfo` also can't use OpenCL is suspicious. I suppose there may be some kind of version mismatch breaking OpenCL.

I'd try creating an issue where there's more technical people able to see the problem.

**Reply by netrunnereve** [reply]
This post popped out at me as I always assumed that WSL straight up doesn't support OpenCL. Apparently it works on Intel GPUs now (see microsoft/WSL/issues/6951), but AFAIK it doesn't work on AMD or NVIDIA.

**Comment by SpeedyCraftah** [comment]
> This post popped out at me as I always assumed that WSL straight up doesn't support OpenCL. Apparently it works on Intel GPUs now (see [microsoft/WSL/issues/6951](https://github.com/microsoft/WSL/issues/6951)), but AFAIK it doesn't work on AMD or NVIDIA.

That's interesting - thank you for bringing it up.
OpenCL seems to suck with compatibility at the moment and I don't think there's much progress on adding Nvidia support to WSL for OpenCL since the issue appears to be stagnant.
I'll go with the native route and try get it running on a real Ubuntu installation.
